### PR TITLE
Simplify PaintCanvas

### PR DESCRIPTION
**Pull request by devildefu**
# Rationale

PaintCanvas is a chimera. It does way more things than it should, which makes it unnecessary complicated and dependent on tokio. 
The point of this PR is to separate: encoding and decoding, loading and saving, to make it simpler and easier to build alternative way to those, to achieve port to browser, or making it synchronous again.

# Changes

## Chunk

Chunk no longer has functions to encode and decode itself. Instead, I moved it to Xcoder.

Also, Chunk is now in separate file, to make `paint_canvas.rs` smaller and nicer.

## Xcoder

A new component has been introduced - Xcoder. It takes care of encoding and decoding chunks, either in synchronous (using `{encode,decode}_{png,webp}_data()`) or asynchronous way (using `enqueue_chunk_{encoding,decoding}()`).

PaintCanvas uses `enqueue_chunk_*` family of functions to delegate encoding or decoding and receives them via channels. It has a nice benefit of making it possible to be synchronous again if needed.

## IoComponent

A new component has been introduced - IoComponent. It takes care of loading and saving canvas. It has two primary functions - `load()` and `save()`, and PaintCanvas is passed to them.

# TLDR

- Move Chunk to separate file
- Encoding and decoding chunks is moved to separate component (Xcoder)
- Loading and saving canvas is moved to separate component (temporarily named IoComponent)

### CODE REVIEW

File: src/app/paint/actions/mod.rs [hunk 1]
```diff
@@ -28,6 +29,7 @@ pub trait Action {
 pub struct ActionArgs<'a> {
    pub assets: &'a Assets,
    pub paint_canvas: &'a mut PaintCanvas,
+   pub iocomponent: &'a mut IoComponent,
```
Comment (liquidev) [inline src/app/paint/actions/mod.rs:32]: When naming variables and modules that contain many words, please separate them with underscores (`iocomponent` should be `io_component`.)

File: src/xcoder.rs [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+
+pub struct XcoderChannels {
+   // Decoder
+   pub decoded_chunks_rx: mpsc::UnboundedReceiver<((i32, i32), RgbaImage)>,
+
+   // Encoder
+   pub encoded_chunks_rx: mpsc::UnboundedReceiver<((i32, i32), ChunkImage)>,
+}
+
+pub struct Xcoder {
```
Comment (liquidev) [inline src/xcoder.rs:21]: Seeing this name in more context now, maybe we should name this `ImageCoder` instead? `Xcoder` is again a name that I feel is too generic for what the component does.

File: src/xcoder.rs [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+
+use crate::chunk::{Chunk, ChunkImage};
+use crate::Error;
+
+pub struct XcoderChannels {
+   // Decoder
+   pub decoded_chunks_rx: mpsc::UnboundedReceiver<((i32, i32), RgbaImage)>,
+
+   // Encoder
+   pub encoded_chunks_rx: mpsc::UnboundedReceiver<((i32, i32), ChunkImage)>,
```
Comment (liquidev) [inline src/xcoder.rs:18]: Any reason to have these comments here? I don't think they add much value on top of the field names already.

File: src/xcoder.rs [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+
+      let decode_join_handle = runtime.spawn(async move {
+         Xcoder::chunk_decoding_loop(
+            runtime2,
+            chunks_to_decode_rx,
+            decoded_chunks_tx,
+            decoder_quit_rx,
+         )
+         .await;
+      });
```
Comment (liquidev) [inline src/xcoder.rs:58]: To avoid the ugly variable name `runtime2`, this can be rewritten using a block expression, like so:
```suggestion
      let decode_join_handle = runtime.spawn({
         let runtime = Arc::clone(&runtime);
         async move {
            Xcoder::chunk_decoding_loop(
               runtime,
               chunks_to_decode_rx,
               decoded_chunks_tx,
               decoder_quit_rx,
            )
            .await;
         }
      });
```

File: src/chunk.rs [hunk 5]
```diff
@@ -0,0 +1,73 @@
+use ::image::{ImageBuffer, Rgba, RgbaImage};
```
Comment (liquidev) [inline src/chunk.rs:1]: Since this file is part of the paint canvas, I think it should belong to the `paint_canvas` module rather than be its own thing. (move this file from `src/chunk.rs` to `src/paint_canvas/chunk.rs` and add `mod chunk;` to `src/paint_canvas.rs` - I know this will be inconsistent with our current style of using `mod.rs` files but we'll get to restructuring the codebase to use the new style some other day)

File: src/iocomponent.rs [hunk 6]
```diff
@@ -0,0 +1,312 @@
...
+
+impl IoComponent {
+   pub fn new(runtime: Arc<Runtime>) -> Self {
+      IoComponent {
+         runtime,
+         filename: None,
+      }
+   }
+
+   /// Saves the entire paint to a PNG file.
```
Comment (liquidev) [inline src/iocomponent.rs:40]: I think there's a spelling mistake here:
```suggestion
   /// Saves the entire paint canvas to a PNG file.
```

File: src/chunk.rs [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+pub struct ChunkImage {
+   pub png: Vec<u8>,
+   pub webp: Option<Vec<u8>>,
+}
+
+/// A chunk on the infinite canvas.
+pub struct Chunk {
+   pub framebuffer: Framebuffer,
+   pub image_cache: Option<ChunkImage>,
+   saved: bool,
```
[the remark this answers is not in the document]
Comment (liquidev) [reply src/chunk.rs]: Alright, so:

- We have the paint canvas, which handles actually painting the pixels to a framebuffer on the GPU, _maybe_ displaying them on the screen. (I'd prefer that to be its own thing but eh whatever, that isn't the point of this PR.)
- Then we'd have a layer on top of that: the image coder. The image coder wraps a paint canvas and gives it image encoding/decoding functionality.
- Then we'd have another layer on top of that: the image cache. It caches encoded images in a map separate from actual chunks.
- And then the newly named `ProjectFile` would be the one that keeps track of which chunks are saved and which are not.

Each layer introduces one bit of functionality on top of the existing chunks. It's a more compositional approach than before. Honestly I'm not sure if there's space for it in this PR since it'll be another large refactoring, so for now we can keep it as is; I was just throwing ideas at the wall here.
